README.md
# Alien Sand

_A double experiment in Canvas pixel-manipulation and Flocking algorithms._

See the boids.js file for the flocking algorithm classes.

See the pixelmap.js file for the pixel-manipulation class.

The aliensand.js is the main controlling program, and show the other classes
in use.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [0.0.2] - 2015-10-17
### Added
- Add basic 'sun' animation

## [0.0.1] - 2015-09-08
### Added
- First iteration

## [0.0.2] - 2015-09-08
### Changed
- Moved the boids and vector code out to separate files
- Updated the README.md file

Copyright 2015, chrisatthestudy <chris@the-study.net>

Copying and distribution of this file, with or without modification, are
permitted provided the copyright notice and this notice are preserved.


src/js/tardis.js
/* 
 *  Danger, Will Robinson -- (sorry, wrong show...) -- crufty code ahead! 
 */
var Tardis = function(context, x, y, scale) {

    var self = {

        setup: function(context, x, y, scale)
        {
            self.context = context;
            self.start_x = x;
            self.start_y = y;
            self.scale = scale;
            self.alpha = 0.0;
        },
    
        update: function()
        {
        },
    
        draw: function(alpha)
        {
            self.alpha = alpha;
            var x = self.start_x;
            var y = self.start_y;
            
            var filldarkblue = 'rgba(0, 0, 128, ' + self.alpha + ')';
            var fillglass = 'rgba(127, 127, 200,' + self.alpha + ')';
            var fillwhite = 'rgba(200, 200, 200, ' + self.alpha + ')';
            var fillblack = 'rgba(0, 0, 0, ' + self.alpha + ')';
            self.context.strokeStyle = fillblack;
    
            var mid = x + (22 * self.scale);
    
            self.context.font = (self.scale * 2) + "px Arial";
    
            self.context.beginPath();
    
            // Light
            w = 2.25 * self.scale;
            h = 0.25 * self.scale;
            x = mid - (w / 2);
            // y = 2 * self.scale;
            self.context.fillStyle = filldarkblue;
            self.context.fillRect(x, y, w, h);
            self.context.rect(x, y, w, h);
            self.context.stroke();
    
            y = y + h;
            w = 2.00 * self.scale;
            h = 2.25 * self.scale;
            x = x + (0.125 * self.scale);
            self.context.fillStyle = fillglass;
            self.context.fillRect(x, y, w, h);
            self.context.rect(x, y, w, h);
            self.context.stroke();
    
            y = y + h;
            w = 2.25 * self.scale;
            h = 0.5 * self.scale;
            x = mid - (w / 2);
            self.context.fillStyle = filldarkblue;
            self.context.fillRect(x, y, w, h);
            self.context.rect(x, y, w, h);
            self.context.stroke();
    
            // Roof
            y = y + h;
            w = 18.25 * self.scale;
            h = 1.5 * self.scale;
            x = mid - (w / 2);
            self.context.fillRect(x, y, w, h);
            self.context.rect(x, y, w, h);
            self.context.stroke();
    
            y = y + h
            w = 22.00 * self.scale;
            h = 1.5 * self.scale;
            x = mid - (w / 2);
            self.context.fillRect(x, y, w, h);
            self.context.rect(x, y, w, h);
            self.context.stroke();
    
            // Main box
            var y1 = y + h
            w = 23.75 * self.scale;
            h = 44.00 * self.scale;
            x = mid - (w / 2);
            self.context.fillRect(x, y1, w, h);
            self.context.rect(x, y1, w, h);
            self.context.stroke();
    
            // Sign box
            y = y1 + (0.5 * self.scale)
            w = 22.00 * self.scale;
            h = 3.00 * self.scale;
            x = mid - (w / 2) - (2 * self.scale);
            self.context.fillRect(x, y, w + (4 * self.scale), h);
            self.context.rect(x, y, w + (4 * self.scale), h);
            self.context.stroke(); 
    
            // Sign
            x = mid - (w / 2);
            var bw = w - ((2.5 * self.scale) * 2);
            var bh = h - ((0.25 * self.scale) * 2);
            self.context.fillStyle = fillblack;
            self.context.fillRect(x + (2.5 * self.scale), y + (0.25 * self.scale), bw, bh); 
    
            self.context.fillStyle = fillwhite;
            self.context.fillText(" POLICE   BOX", x + (3.0 * self.scale), y + (2.25 * self.scale));
    
            self.context.fillStyle = filldarkblue;
            var doorTop = y + h + (1.5 * self.scale);
    
            var door = function()
            {
                w = 9.00 * self.scale;
                h = 38.00 * self.scale;
                self.context.fillStyle = filldarkblue;
                self.context.fillRect(x, y, w, h);
                self.context.rect(x, y, w, h);
                self.context.stroke();
        
                // Window Panel
                y = y + (1.5 * self.scale);
                x = x + (1.5 * self.scale);
                w = 6.0 * self.scale;
                h = 7.75 * self.scale;
                self.context.fillStyle = fillglass;
                self.context.fillRect(x, y, w, h);
                self.context.rect(x, y, w, h);
                self.context.stroke();
        
                // Window cross-pieces
                self.context.beginPath();
                var ly = y + 1;
                var lx = x + (2.0 * self.scale);
                var lh = h - 2;
                var lw = w - 2;
        
                // -- vertical
                self.context.moveTo(lx, ly);
                self.context.lineTo(lx, ly + lh);
        
                lx = lx + (2.0 * self.scale);
                self.context.moveTo(lx, ly);
                self.context.lineTo(lx, ly + lh);
        
                // -- horizontal
                lx = x + 1;
                ly = ly + (2.0 * self.scale);
                self.context.moveTo(lx, ly);
                self.context.lineTo(lx + lw, ly);
        
                ly = ly + (3.0 * self.scale);
                self.context.moveTo(lx, ly);
                self.context.lineTo(lx + lw, ly);
        
                self.context.strokeStyle = fillwhite;
                self.context.lineWidth = 0.5 * self.scale;
                self.context.stroke();
        
                self.context.beginPath();
                self.context.lineWidth = (1.0 * self.scale);
                self.context.strokeStyle = fillblack;
        
                // Inset panels
                for (var i = 0; i < 3; i++)
                {
                    y = y + h + (1.5 * self.scale);
                    self.context.fillStyle = filldarkblue;
                    self.context.fillRect(x, y, w, h);
                    self.context.rect(x, y, w, h);
                    self.context.stroke();
                }
            }
    
            // Left door
            x = mid - (9.00 * self.scale);
            y = doorTop;
            door(mid - (9.00 * self.scale), doorTop);
    
            // Right door
            x = mid;
            y = doorTop;
            door(mid, doorTop);
    
            // Base
            y = doorTop + (39.0 * self.scale);
            w = 27.0 * self.scale;
            h = 0.75 * self.scale;
            x = mid - (w / 2);
            self.context.fillRect(x, y, w, h);
            self.context.rect(x, y, w, h);
            self.context.stroke();
        }
    };
    self.setup(context, x, y, scale);
    return self;
}


src/js/pixelmap.js
/*
 *  ============================================================================
 *  PixelMap()
 *  ============================================================================
 *  Object for handling the direct pixel manipulation of the canvas. This takes
 *  a section of the canvas (defined by x, y, w, h) and allows the pixels
 *  within it to be set or cleared. Pixel positions within the PixelMap are
 *  defined by the u, v variables.
 *  ----------------------------------------------------------------------------
 */
//{{{
var PixelMap = function( options ) {
    'use strict';
    
    // The actual object representing this function is assigned to 'self', 
    // which is returned from this function so that the returned value is
    // effectively an alias for PixelMap.self.
    var self = {
        
        // setup() performs all the necessary initialisation
        setup: function( options ) {
            debug("PixelMap.setup()", 1);
            try {
                self.x = options.x || 0;
                self.y = options.y || 0;
                self.w = options.w || 8;
                self.h = options.h || 8;
                
                self.context = options.context;

                // Create an internal imageData object  
                self.frame = self.context.createImageData(self.w, self.h);
                self.length = self.frame.data.length / 4;
                debug("PixelMap.length=" + self.length, 1);
            }
            catch(e) {
                debug("PixelMap.setup error: " + e.Message, 1);
            }
        },
        
        update: function( ) {
            debug("PixelMap.update()");
        },

        set: function(u, v, color) {
            try {
                var idx = ((v * self.w) + u) * 4;
                self.frame.data[idx] = color.r;
                self.frame.data[idx + 1] = color.g;
                self.frame.data[idx + 2] = color.b;
                self.frame.data[idx + 3] = color.a || 255;
            }
            catch(e) {
                debug("PixelMap.set error: " + e.Message, 1);
            }
        },
        
        get: function(u, v) {
            try {
                var idx = ((v * self.w) + u) * 4;
                var r = self.frame.data[idx];
                var g = self.frame.data[idx + 1];
                var b = self.frame.data[idx + 2];
                var a = self.frame.data[idx + 3];
                return { r: r, g: g, b: b, a: a };
            }
            catch(e) {
                debug("PixelMap.get error: " + e.Message, 1);
            }
        },
        
        clear: function( color ) {
            try {
                var idx;
                var i;
                for (i = 0; i < self.length; i++) {
                    idx = i * 4;
                    this.frame.data[idx] = color.r;
                    this.frame.data[idx + 1] = color.g;
                    this.frame.data[idx + 2] = color.b;
                    this.frame.data[idx + 3] = color.a;
                }
            }
            catch(e) {
                debug("PixelMap.clear error: " + e.Message, 1);
            }
        },
        
        // draw() renders the object on-screen
        draw: function( ) {
            debug("PixelMap.draw()");
            try {
                // Copy the pixel map to the main canvas  
                self.context.putImageData(self.frame, self.x, self.y);
            }
            catch(e) {
                debug("PixelMap.draw error: " + e.Message, 1);
            }
        }
    };

    // Call our setup() function, passing on any options that were passed to
    // the function
    self.setup( options );
    return self;
};
//}}}


src/js/aliensand.js
/*jslint browser: true*/
/*jslint plusplus: true */
/*global window, console */

var AlienSand = (function() {
    
    /*
     *  ============================================================================
     *  requestAnim
     *  ============================================================================
     *  Shim layer by Paul Irish.
     *  ----------------------------------------------------------------------------
     */
    window.requestAnimFrame = (function(){
      'use strict';
      return  window.requestAnimationFrame       ||
              window.webkitRequestAnimationFrame ||
              window.mozRequestAnimationFrame    ||
              window.oRequestAnimationFrame      ||
              window.msRequestAnimationFrame     ||
              function(/* function */ callback, /* DOMElement */ element){
                  window.setTimeout(callback, 16.666);
              };
    }());

    /*
     *  ============================================================================
     *  Sun()
     *  ============================================================================
     *  Simple representation of the sun.
     *  ----------------------------------------------------------------------------
     */
    var Sun = function( options ) {

	var self = {

	    setup: function( options ) {
		self.context = options.context;
		self.angle_units = (2 * Math.PI) / 6400.0;
		self.angle = (self.angle_units * 6400.0);
		self.halo_step = 1;
	    },

	    update: function( ) {
	    },

	    draw: function( ) {
		// x(t) = r cos(t) + j       y(t) = r sin(t) + k
		var x = (400 * Math.cos(self.angle)) + 150;
		var y = (800 * Math.sin(self.angle)) + 250;
                self.context.fillStyle = "#004400";
		self.context.beginPath();
		self.context.arc(x, y, 20, 0, Math.PI * 2, false);
		self.context.fill();
		
		self.context.beginPath();
		self.context.lineWidth = 10;
                self.context.strokeStyle = 'rgba(0,64,0,' + (1.0 - self.halo_step / 10) + ')';
		console.log(self.context.strokeStyle);
		self.context.arc(x, y, 20 + self.halo_step, 0, Math.PI * 2, false);
		self.context.stroke();
		
		self.halo_step = self.halo_step + 0.1;
		if (self.halo_step > 10) {
		    self.halo_step = 1;
		}
		self.angle = self.angle - self.angle_units;
		if (self.angle < 0) {
		    self.angle = (self.angle_units * 6400.0);
		}
		self.context.lineWidth = 1;
	    }
	};
	self.setup( options );
	return self;
    };
    
    /*
     *  ============================================================================
     *  App()
     *  ============================================================================
     *  The main application class. Call 'execute' on this to start it running.
     *  ----------------------------------------------------------------------------
     */
    var App = function( options ) {
        'use strict';
        
        // The actual object representing this function is assigned to 'self', 
        // which is return from this function so that App is effectively an alias
        // for App.self.
        var self = {
            
            // setup() performs all the necessary initialisation
            setup: function( options ) {
                // Get the canvas and context and store these away
                self.canvas = document.getElementById('canvas');
                self.context = canvas.getContext('2d');
    
                self.startTime = (new Date()).getTime();
                self.prevTime  = self.startTime;
                self.fps       = 0;
                self.ticks     = 0;
                self.tardis_frame = 0;
                self.tardis_fade = 0;

                var scale = 3.0;
                self.tardis = Tardis(self.context, self.canvas.width - 150, self.canvas.height - (40 + (scale * 51)), scale)
		self.sun = Sun( { context: self.context } );
                
                self.running = true;
                
                self.flock = Flock( { context: self.context } );
                
            },
            
            // update() is called on each frame tick
            update: function( ) {
                var currentTime;
                var delta = (currentTime=new Date()) - self.prevTime;
                self.fps += (delta - self.fps) / 30;
                self.prevTime = currentTime;
                self.ticks += delta;
                
                if (self.ticks > 120) {
                    self.tardis_frame += 1;
                    if (self.tardis_frame > 250) {
                        self.tardis_fade -= 0.05;
                        if (self.tardis_fade < 0.0) {
                            self.tardis_fade = 0;
                            self.tardis_frame = 0;
                        }
                    } else if (self.tardis_frame > 100) {
                        self.tardis_fade += 0.05;
                        if (self.tardis_fade > 255.0) {
                            self.tardis_fade = 255.0;
                        }
                    }
                    self.flock.update( );
		    self.sun.update( );
                    self.ticks = 0;
                }
            },
            
            draw: function( ) {
                // Temporary hack for mobile displays
                if (window.innerWidth < 481) {
                    self.canvas.width  = window.innerWidth;
                }
                // Draw the 'ground'
                self.context.fillStyle = "#222922";
                self.context.fillRect(0, 0, self.canvas.width, self.canvas.height);

		self.sun.draw( );
		
                self.context.fillStyle = "#004400";
                self.context.fillRect(0, self.canvas.height - 50, self.canvas.width, 50);
                // Draw the boids
                self.flock.draw( );
                // Draw the TARDIS
                var scale = 3.0;
                self.tardis.start_x = self.canvas.width - 150;
                self.tardis.start_y = self.canvas.height - (40 + (scale * 51))
                self.tardis.draw( self.tardis_fade );
            },
            
            // execute() runs the main loop
            execute: function( ) {
                debug("Updating");
                // Update all the components
                self.update( );
                
                // Clear the screen
                self.context.clearRect(0, 0, self.canvas.width, self.canvas.height);
    
                debug("Drawing");
                // Draw all the components
                self.draw( );
                
                // Recursively call this function for as long as we are running
                if (self.running) {
                    requestAnimFrame( self.execute );
                }
            },
            
        };
    
        // Call our setup() function, passing on any options that were passed to
        // the App function
        self.setup( options );
        return self;
    };

    function hideNoJS()
    {
        var elementArray = document.getElementsByClassName("no-js");
    
        for(var i = (elementArray.length - 1); i >= 0; i--)
        {
            elementArray[i].className = "hide-me";
        }
    }
    
    /*
     *  ============================================================================
     *  document.onreadystatechange
     *  ============================================================================
     *  The document.onreadystatechange event is called by the browser while it
     *  is loading the current document. Once the document.readyState is 'complete',
     *  the document has finished loading, and we can create and start the app.
     *  ----------------------------------------------------------------------------
     */
    document.onreadystatechange = function () {
        // console.log(document.readyState);
        if (document.readyState == "complete") {
            debug("Document ready -- creating App");
            hideNoJS();
            var app = App();
            debug("Executing App");
            app.execute();
            //start();
        }
    };

})()    



src/js/boids.js
/*
 *  ============================================================================
 *  Vector3d
 *  ============================================================================
 *  Simple vector class. It can also be used for 2d, by leaving the z parameter
 *  at 0.
 *  ----------------------------------------------------------------------------
 */
var Vector3d = function( x, y, z ) {
    'use strict';

    var self = {
	// setup() - simple initialisation, called automatically when the
	// function is invoked
	setup: function( x, y, z ) {
	    self.x = x || 0;
	    self.y = y || 0;
	    self.z = z || 0;
	},

	// copy() - returns a copy of this vector
	copy: function( ) {
	    var result = Vector3d();
	    result.x = self.x;
	    result.y = self.y;
	    result.z = self.z;
	    return result;
	},

	// add() - adds the supplied vector, which should support at least the
	// x and y fields, and optionally a z field (this will default to 0 if
	// it does not exist). Normally this will be another Vector3d.
	// It returns the new vector
	add: function( vector ) {
	    var result = self.copy();
	    result.x += vector.x;
	    result.y += vector.y;
	    result.z += vector.z || 0;
	    return result;
	},

	// subtract() - subtracts the supplied vector, using the same rules
	// and limitations of the add() function. It returns the new vector.
	subtract: function( vector ) {
	    var result = self.copy();
	    result.x -= vector.x;
	    result.y -= vector.y;
	    result.z -= vector.z || 0;
	    return result;
	},

	// multiply() - scalar multiplication by the supplied multiplicand.
	multiply: function( multiplicand ) {
	    var result = self.copy();
	    result.x *= multiplicand;
	    result.y *= multiplicand;
	    result.z *= multiplicand;
	    return result;
	},

	// divide() - scalar division by the supplied dividend. Ignores any
	// attempt to divide by zero. It returns the new vector;
	divide: function( dividend ) {
	    var result = self.copy();
	    if (dividend !== 0) {
		result.x /= dividend;
		result.y /= dividend;
		result.z /= dividend;
	    }
	    return result;
	},

	// displacement() - returns the distance from this vector to the
	// supplied vector. This will always be a positive value.
	displacement: function( fromVector ) {
	    var vector = Vector3d;
	    vector = self.copy();
	    vector = fromVector.subtract(vector);
	    return Math.sqrt(Math.pow(vector.x, 2) + Math.pow(vector.y, 2) + Math.pow(vector.z, 2));
	},

	// magnitude() - returns the length of the vector
	magnitude: function( ) {
	    return Math.sqrt(Math.pow(self.x, 2) + Math.pow(self.y, 2) + Math.pow(self.z, 2));
	}
    };

    self.setup( x, y, z );
    return self;
};

/*
 *  ============================================================================
 *  Boid()
 *  ============================================================================
 *  Handles a single Boid
 *  ----------------------------------------------------------------------------
 */
var Boid = function( options ) {
    'use strict';

    // The actual object representing this function is assigned to 'self', 
    // which is returned from this function so that the returned value 
    // effectively an alias for Boid.self.
    var self = {

	// setup() performs all the necessary initialisation
	setup: function( options ) {
	    debug("setup()");

	    self.options = options;
	    self.flock = options.flock;

	    // Add the sprite set-up details to the options
	    self.options.w = 3;
	    self.options.h = 3;

	    self.sprite = PixelMap( self.options );

	    self.position = Vector3d( self.options.x, self.options.y );
	    self.velocity = Vector3d( Math.random() * 3, Math.random() * 3 );

	    // Set the wing-position (this actually maps to the position of
	    // the pixels that represent the wings: 1 = up, 2 = middle,
	    // 3 = down). Select a random position.
	    self.wingpos = Math.floor((Math.random() * 3)) + 1;
	    if (self.wingpos == 3) {
		self.wingdir = -1;
	    }
	    else if (self.wingpos == 1) {
		self.wingdir = 1;
	    }
	    else {
		self.wingdir = 1;
	    }

	},

	// update() is called on each tick
	update: function( ) {
	    debug("update()");
	    var color = {r: 0, g:250, b:0, a: 0};

	    // Move the wing
	    self.wingpos = self.wingpos + self.wingdir;
	    if ((self.wingpos == 3) || (self.wingpos == 1)) {
		self.wingdir = self.wingdir * -1;
	    }

	    // Clear to transparent
	    self.sprite.clear( color );

	    // Set to opaque, and draw the boid
	    color.a = 255;
	    self.sprite.set(0, self.wingpos - 1, color);
	    self.sprite.set(1, 1, color);
	    self.sprite.set(2, self.wingpos - 1, color);

	    // Apply the flocking rules
	    var v1 = self.moveTowardsCentre();
	    var v2 = self.maintainDistance();
	    var v3 = self.matchVelocity();
	    var v4 = self.keepInBounds();

	    self.velocity = self.velocity.add(v1);
	    self.velocity = self.velocity.add(v2);
	    self.velocity = self.velocity.add(v3);
	    self.velocity = self.velocity.add(v4);

	    debug("Changing position", 2);
	    self.limitVelocity();
	    self.position = self.position.add(self.velocity);
	},

	// moveTowardsCentre() moves the boid towards the centre of the
	// flock
	moveTowardsCentre: function( ) {
	    debug("moveTowardsCentre", 2);
	    var centre = Vector3d();
	    var count = self.flock.count();
	    var boid = null;
	    for (var i = 0; i < count; i++) {
		boid = self.flock.boids[i]; 
		if (boid != self) {
		    centre = centre.add(boid.position);                    
		}
	    }
	    // Final calculation of the centre
	    centre = centre.divide( count );

	    // How far to move the boid towards the centre
	    centre = centre.subtract(self.position);
	    centre = centre.divide( 100 );

	    // Return the vector representing how far the boid should move
	    return centre.copy();
	},

	// maintainDistance() keeps the boid a certain distance away from any
	// other boids
	maintainDistance: function( ) {
	    debug("maintainDistance", 2);
	    var distance = Vector3d();
	    var displacement;
	    var offset;
	    var boid;
	    var count = self.flock.count();
	    for (var i = 0; i < count; i++) {
		boid = self.flock.boids[i]; 
		if (boid != self) {
		    displacement = self.position.displacement(boid.position);
		    if (displacement < 5) {
			offset = self.position.subtract(boid.position);
			distance = distance.subtract(offset);
		    }                        
		}
	    }
	    return distance.copy();
	},

	// matchVelocity() keeps the boid's velocity matched with the velocity
	// of nearby boids
	matchVelocity: function( ) {
	    debug("matchVelocity", 2);
	    var boid;
	    var result = Vector3d();
	    var count = self.flock.count();
	    for (var i = 0; i < count; i++) {
		boid = self.flock.boids[i]; 
		if (boid != self) {
		    result = result.add(boid.velocity);
		}
	    }
	    result = result.divide( count );
	    result = result.subtract(self.velocity);
	    result = result.divide( 8 );
	    return result.copy();
	},

	// keepInBounds() keeps the boid within the canvas area
	keepInBounds: function( ) {
	    var result = Vector3d();
	    if (self.position.x < 50) {
		result.x = 10;
	    } else if (self.position.x > (self.options.context.canvas.width - 50)) {
		result.x = -10;
	    }
	    if (self.position.y < 50) {
		result.y = 10;
	    } else if (self.position.y > (self.options.context.canvas.height - 50)) {
		result.y = -10;
	    }
	    return result;
	},

	// limitVelocity() prevents the boid from moving too quickly, which
	// can happen as a result of the accumulation of the other rules. If
	// you remove the call to this function, the boids will get faster
	// and faster over time.
	limitVelocity: function( ) {
	    var limit = 20;
	    var speed = self.velocity.magnitude( );
	    if (speed > limit) {
		self.velocity = self.velocity.divide(limit);
	    }
	},

	draw: function( ) {
	    debug("draw()");
	    self.sprite.x = self.position.x;
	    self.sprite.y = self.position.y;
	    self.sprite.draw( );
	}
    };

    // Call our setup() function, passing on any options that were passed to
    // the function
    self.setup( options );
    return self;
};

/*
 *  ============================================================================
 *  Flock()
 *  ============================================================================
 *  Models a flock of boids
 *  ----------------------------------------------------------------------------
 */
var Flock = function( options ) {
    'use strict';

    // The actual object representing this function is assigned to 'self', 
    // which is returned from this function so that the returned value 
    // effectively an alias for Flock.self.
    var self = {

	// setup() performs all the necessary initialisation
	setup: function( options ) {
	    debug("setup()");
	    var i = 0;

	    // Set up the default values for the boids
	    var boid_options = { flock: self, context: options.context };

	    // Create the boids
	    self.boids = [];
	    var count = 50;
	    for (i = 0; i < count; i++) {
		// Select a random starting position
		boid_options.x = Math.random() * 100;
		boid_options.y = Math.random() * 100;
		self.boids.push(Boid(boid_options));
	    }
	},

	// update() is called on each tick
	update: function( ) {
	    debug("update()");
	    // Update each boid
	    for (var i = 0; i < self.boids.length; i++) {
		self.boids[i].update();
	    }
	},

	draw: function( ) {
	    debug("draw()");
	    // Draw all the boids
	    for (var i = 0; i < self.boids.length; i++) {
		self.boids[i].draw();
	    }
	},

	count: function( ) {
	    return self.boids.length;
	}
    };

    // Call our setup() function, passing on any options that were passed to
    // the function
    self.setup( options );
    return self;
};
